src/storage/txn/tablet_flow_controller.rs

```diff
@@ -0,0 +1,727 @@
+                        }
+                    }
+                }
+                tikv_alloc::remove_thread_memory_accessor();
+            })
+            .unwrap()
+    }
+}
+
+impl FlowController for TabletFlowController {
+    fn should_drop(&self, region_id: u64) -> bool {
+        let limiters = self.limiters.as_ref().read().unwrap();
+        if let Some(limiter) = limiters.get(&region_id) {
+            let ratio = limiter.1.load(Ordering::Relaxed);
+            let mut rng = rand::thread_rng();
+            return rng.gen_ratio(ratio, RATIO_SCALE_FACTOR);
+        }
+        false
+    }
+
+    #[cfg(test)]
+    fn discard_ratio(&self, region_id: u64) -> f64 {
+        let limiters = self.limiters.as_ref().read().unwrap();
+        if let Some(limiter) = limiters.get(&region_id) {
+            let ratio = limiter.1.load(Ordering::Relaxed);
+            return ratio as f64 / RATIO_SCALE_FACTOR as f64;
+        }
+        0.0
+    }
+
+    fn consume(&self, region_id: u64, bytes: usize) -> Duration {
+        let limiters = self.limiters.as_ref().read().unwrap();
+        if let Some(limiter) = limiters.get(&region_id) {
+            return limiter.0.consume_duration(bytes);
+        }
+        Duration::ZERO
+    }
+
+    fn unconsume(&self, region_id: u64, bytes: usize) {
+        let limiters = self.limiters.as_ref().read().unwrap();
+        if let Some(limiter) = limiters.get(&region_id) {
+            limiter.0.unconsume(bytes);
+        }
+    }
+
+    #[cfg(test)]
+    fn total_bytes_consumed(&self, region_id: u64) -> usize {
+        let limiters = self.limiters.as_ref().read().unwrap();
+        if let Some(limiter) = limiters.get(&region_id) {
+            return limiter.0.total_bytes_consumed();
+        }
+        0
+    }
+
+    fn enable(&self, enable: bool) {
+        self.enabled.store(enable, Ordering::Relaxed);
+        if let Some(tx) = &self.tx {
+            if enable {
+                tx.send(Msg::Enable).unwrap();
+            } else {
+                tx.send(Msg::Disable).unwrap();
+            }
+        }
+    }
+
+    fn enabled(&self) -> bool {
+        self.enabled.load(Ordering::Relaxed)
+    }
+
+    #[cfg(test)]
+    fn set_speed_limit(&self, region_id: u64, speed_limit: f64) {
+        let limiters = self.limiters.as_ref().read().unwrap();
+        if let Some(limiter) = limiters.get(&region_id) {
+            limiter.0.set_speed_limit(speed_limit);
+        }
+    }
+
+    fn is_unlimited(&self, region_id: u64) -> bool {
+        let limiters = self.limiters.as_ref().read().unwrap();
+        if let Some(limiter) = limiters.get(&region_id) {
+            return limiter.0.speed_limit() == f64::INFINITY;
+        }
+        true
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use engine_rocks::FlowInfo;
+    use engine_traits::DummyFactory;
+
+    use super::{
+        super::flow_controller::{tests::*, *},
+        *,
+    };
+
+    fn test_tablet_flow_controller()
```

Comment:
should rename to `make_test_tablet_flowxxx` or similar

Reply:
good point.